README.md
[![Build Status](https://travis-ci.org/Azure/doAzureParallel.svg?branch=master)](https://travis-ci.org/Azure/doAzureParallel)
# doAzureParallel

## Introduction

The *doAzureParallel* package is a parallel backend for the widely popular *foreach* package. With *doAzureParallel*, each iteration of the *foreach* loop runs in parallel on an Azure Virtual Machine (VM), allowing users to scale up their R jobs to tens or hundreds of machines.

*doAzureParallel* is built to support the *foreach* parallel computing package. The *foreach* package supports parallel execution - it can execute multiple processes across some parallel backend. With just a few lines of code, the *doAzureParallel* package helps create a cluster in Azure, register it as a parallel backend, and seamlessly connects to the *foreach* package.

NOTE: The terms *pool* and *cluster* are used interchangably throughout this document.

## Notable Features
- Ability to use low-priority VMs for an 80% discount [(link)](./docs/31-vm-sizes.md#low-priority-vms)
- Users can bring their own Docker Image
- AAD and VNets Support
- Built in support for Azure Blob Storage

## Dependencies

- R (>= 3.3.1)
- httr (>= 1.2.1)
- rjson (>= 0.2.15)
- RCurl (>= 1.95-4.8)
- digest (>= 0.6.9)
- foreach (>= 1.4.3)
- iterators (>= 1.0.8)
- bitops (>= 1.0.5)

## Setup 

1) Install doAzureParallel directly from Github.

```R
# install the package devtools
install.packages("devtools")

# install the doAzureParallel and rAzureBatch package
devtools::install_github("Azure/rAzureBatch")
devtools::install_github("Azure/doAzureParallel")
```

2) Create an doAzureParallel's credentials file
``` R
library(doAzureParallel)
generateCredentialsConfig("credentials.json")
```

3) Login or register for an Azure Account, navigate to [Azure Cloud Shell](https://shell.azure.com)

``` sh 
wget -q https://raw.githubusercontent.com/Azure/doAzureParallel/master/account_setup.sh &&
chmod 755 account_setup.sh &&
/bin/bash account_setup.sh
```
4) Follow the on screen prompts to create the necessary Azure resources and copy the output into your credentials file. For more information, see [Getting Started Scripts](./docs/02-getting-started-script.md).

To Learn More:
- [Azure Account Requirements for doAzureParallel](./docs/04-azure-requirements.md)

## Getting Started

Import the package
```R
library(doAzureParallel)
```

Set up your parallel backend with Azure. This is your set of Azure VMs.
```R
# 1. Generate your credential and cluster configuration files.  
generateClusterConfig("cluster.json")
generateCredentialsConfig("credentials.json")

# 2. Fill out your credential config and cluster config files.
# Enter your Azure Batch Account & Azure Storage keys/account-info into your credential config ("credentials.json") and configure your cluster in your cluster config ("cluster.json")

# 3. Set your credentials - you need to give the R session your credentials to interact with Azure
setCredentials("credentials.json")

# 4. Register the pool. This will create a new pool if your pool hasn't already been provisioned.
cluster <- makeCluster("cluster.json")

# 5. Register the pool as your parallel backend
registerDoAzureParallel(cluster)

# 6. Check that your parallel backend has been registered
getDoParWorkers()
```

Run your parallel *foreach* loop with the *%dopar%* keyword. The *foreach* function will return the results of your parallel code.

```R
number_of_iterations <- 10
results <- foreach(i = 1:number_of_iterations) %dopar% {
  # This code is executed, in parallel, across your cluster.
  myAlgorithm()
}
```

After you finish running your R code in Azure, you may want to shut down your cluster of VMs to make sure that you are not being charged anymore.

```R
# shut down your pool
stopCluster(cluster)
```

## Table of Contents 
This section will provide information about how Azure works, how best to take advantage of Azure, and best practices when using the doAzureParallel package.

1. **Azure Introduction** [(link)](./docs/00-azure-introduction.md)

   Using *Azure Batch*

2. **Getting Started** [(link)](./docs/01-getting-started.md)

    Using the *Getting Started* to create credentials
    
    i. **Generate Credentials Script** [(link)](./docs/02-getting-started-script.md)

    - Pre-built bash script for getting Azure credentials without Azure Portal

    ii. **National Cloud Support** [(link)](./docs/03-national-clouds.md)

    - How to run workload in Azure national clouds

3. **Customize Cluster** [(link)](./docs/30-customize-cluster.md)

    Setting up your cluster to user's specific needs

    i. **Virtual Machine Sizes** [(link)](./docs/31-vm-sizes.md)
    
    - How do you choose the best VM type/size for your workload?

    ii. **Autoscale** [(link)](./docs/32-autoscale.md)
  
    - Automatically scale up/down your cluster to save time and/or money.
  
    iii. **Building Containers** [(link)](./docs/33-building-containers.md)
    
      - Creating your own Docker containers for reproducibility

4. **Managing Cluster** [(link)](./docs/40-clusters.md)

    Managing your cluster's lifespan

5. **Customize Job**

    Setting up your job to user's specific needs
    
    i. **Asynchronous Jobs** [(link)](./docs/51-long-running-job.md)
    
    - Best practices for managing long running jobs
  
    ii. **Foreach Azure Options** [(link)](./docs/52-azure-foreach-options.md)
        
    - Use Azure package-defined foreach options to improve performance and user experience
  
    iii. **Error Handling** [(link)](./docs/53-error-handling.md)
    
    - How Azure handles errors in your Foreach loop? 
    
6. **Package Management** [(link)](./docs/20-package-management.md)

    Best practices for managing your R packages in code. This includes installation at the cluster or job level as well as how to use different package providers.

7. **Storage Management**
    
    i. **Distributing your Data** [(link)](./docs/71-distributing-data.md)
    
    - Best practices and limitations for working with distributed data.

    ii. **Persistent Storage** [(link)](./docs/72-persistent-storage.md)

    - Taking advantage of persistent storage for long-running jobs
   
    iii. **Accessing Azure Storage through R** [(link)](./docs/73-managing-storage.md)
    
    - Manage your Azure Storage files via R 

8. **Performance Tuning** [(link)](./docs/80-performance-tuning.md)

    Best practices on optimizing your Foreach loop

9. **Debugging and Troubleshooting** [(link)](./docs/90-troubleshooting.md)
    
    Best practices on diagnosing common issues

10. **Azure Limitations** [(link)](./docs/91-quota-limitations.md)

    Learn about the limitations around the size of your cluster and the number of foreach jobs you can run in Azure.
   
## Additional Documentation
Read our [**FAQ**](./docs/92-faq.md) for known issues and common questions.

## Next Steps

For more information, please visit [our documentation](./docs/README.md).


CHANGELOG.md
## 0.8.0 (2019-06-19)

* Feature: Upgrading to R Batch SDK to 2018-12-01.8.0 (#354) ([96bfc22](https://github.com/Azure/doAzureParallel/commit/96bfc22)), closes [#354](https://github.com/Azure/doAzureParallel/issues/354)
* Fix: Remove 'List Task Counts' api dependency for job progress (#354) ([96bfc22](https://github.com/Azure/doAzureParallel/commit/96bfc22)), closes [#354](https://github.com/Azure/doAzureParallel/issues/354)


## <small>0.7.3 (2019-04-16)</small>

* Added documentation for containerImage (#339) ([82c149a](https://github.com/Azure/doAzureParallel/commit/82c149a)), closes [#339](https://github.com/Azure/doAzureParallel/issues/339)
* Added fix for cluster of autoscale node 0 (#337) ([40b74da](https://github.com/Azure/doAzureParallel/commit/40b74da)), closes [#337](https://github.com/Azure/doAzureParallel/issues/337)
* Minor documentation  fixes (#331) ([6d14d45](https://github.com/Azure/doAzureParallel/commit/6d14d45)), closes [#331](https://github.com/Azure/doAzureParallel/issues/331)
* Update pipeline.yml for Azure Pipelines ([a2154d2](https://github.com/Azure/doAzureParallel/commit/a2154d2))
* Fix: Remove unnecessary apt packages installation on job preparation task (#321) ([975d68e](https://github.com/Azure/doAzureParallel/commit/975d68e)), closes [#321](https://github.com/Azure/doAzureParallel/issues/321)
* Fix: Revert Cluster Setup File (#342) ([372e382](https://github.com/Azure/doAzureParallel/commit/372e382)), closes [#342](https://github.com/Azure/doAzureParallel/issues/342)
* Fix: Test Coverage on  Azure Pipeline CI (#348) ([9758580](https://github.com/Azure/doAzureParallel/commit/9758580)), closes [#348](https://github.com/Azure/doAzureParallel/issues/348)
* Bugfix: Don't break when fetching batch-insights (#340) ([ec1a4df](https://github.com/Azure/doAzureParallel/commit/ec1a4df)), closes [#340](https://github.com/Azure/doAzureParallel/issues/340)
* Feature: Added Queue/Running Autoscale Formula (#316) ([7192e08](https://github.com/Azure/doAzureParallel/commit/7192e08)), closes [#316](https://github.com/Azure/doAzureParallel/issues/316)
* Feature: Batch Insights support (#336) ([31eec56](https://github.com/Azure/doAzureParallel/commit/31eec56)), closes [#336](https://github.com/Azure/doAzureParallel/issues/336)



## <small>0.7.2 (2018-10-08)</small>

* Release: v0.7.2 (#317) ([844ebde](https://github.com/Azure/doAzureParallel/commit/844ebde)), closes [#317](https://github.com/Azure/doAzureParallel/issues/317)
* Fix: Documentation typos and fixes (#307) ([1d26dd8](https://github.com/Azure/doAzureParallel/commit/1d26dd8)), closes [#307](https://github.com/Azure/doAzureParallel/issues/307)
* Fix: File share sample (#312) ([96081cc](https://github.com/Azure/doAzureParallel/commit/96081cc)), closes [#312](https://github.com/Azure/doAzureParallel/issues/312)
* Added 0.7.1 changelog version (#286) ([ad7acf1](https://github.com/Azure/doAzureParallel/commit/ad7acf1)), closes [#286](https://github.com/Azure/doAzureParallel/issues/286)
* Added docs (#298) ([b9ffe3e](https://github.com/Azure/doAzureParallel/commit/b9ffe3e)), closes [#298](https://github.com/Azure/doAzureParallel/issues/298)
* Added github branch installation (#302) ([4020a93](https://github.com/Azure/doAzureParallel/commit/4020a93)), closes [#302](https://github.com/Azure/doAzureParallel/issues/302)
* Fix Azure foreach options documentation (#288) ([ce263e7](https://github.com/Azure/doAzureParallel/commit/ce263e7)), closes [#288](https://github.com/Azure/doAzureParallel/issues/288)
* Fix dependency on enum being uppercase (#310) ([81dcb72](https://github.com/Azure/doAzureParallel/commit/81dcb72)), closes [#310](https://github.com/Azure/doAzureParallel/issues/310)
* Feature: Add Batch Service Built-In Container support (#291) ([b97f447](https://github.com/Azure/doAzureParallel/commit/b97f447)), closes [#291](https://github.com/Azure/doAzureParallel/issues/291)



## <small>0.7.1 (2018-07-13)</small>

* add sharedKey to credentials related code and doc (#266) ([6582af4](https://github.com/Azure/doAzureParallel/commit/6582af4)), closes [#266](https://github.com/Azure/doAzureParallel/issues/266)
* CHANGELOG for v0.7.1 (#285) ([737bf49](https://github.com/Azure/doAzureParallel/commit/737bf49)), closes [#285](https://github.com/Azure/doAzureParallel/issues/285)
* Clarifying comment on DSVM (#274) ([008b0ad](https://github.com/Azure/doAzureParallel/commit/008b0ad)), closes [#274](https://github.com/Azure/doAzureParallel/issues/274)
* Documentation rewrite (#273) ([f418cd9](https://github.com/Azure/doAzureParallel/commit/f418cd9)), closes [#273](https://github.com/Azure/doAzureParallel/issues/273)
* Feature/custom package (#272) ([20c86f1](https://github.com/Azure/doAzureParallel/commit/20c86f1)), closes [#272](https://github.com/Azure/doAzureParallel/issues/272)
* Fix/getstartdoc (#263) ([81a7c16](https://github.com/Azure/doAzureParallel/commit/81a7c16)), closes [#263](https://github.com/Azure/doAzureParallel/issues/263)
* Fix/storage management (#267) ([84aa7c9](https://github.com/Azure/doAzureParallel/commit/84aa7c9)), closes [#267](https://github.com/Azure/doAzureParallel/issues/267)
* Fixed output (#270) ([d02599d](https://github.com/Azure/doAzureParallel/commit/d02599d)), closes [#270](https://github.com/Azure/doAzureParallel/issues/270)
* Fixed package strip name (#278) ([f35e1e3](https://github.com/Azure/doAzureParallel/commit/f35e1e3)), closes [#278](https://github.com/Azure/doAzureParallel/issues/278)
* Replaced deprecated function (#269) ([23079f5](https://github.com/Azure/doAzureParallel/commit/23079f5)), closes [#269](https://github.com/Azure/doAzureParallel/issues/269)
* Switched to R6 sas client (#276) ([b24f20c](https://github.com/Azure/doAzureParallel/commit/b24f20c)), closes [#276](https://github.com/Azure/doAzureParallel/issues/276)
* Tests/r (#275) ([4983fb1](https://github.com/Azure/doAzureParallel/commit/4983fb1)), closes [#275](https://github.com/Azure/doAzureParallel/issues/275)

R/utility.R
#' Polling method to check status of cluster boot up
#'
#' @param poolId The cluster name to poll for
#' @param timeout Timeout in seconds, default timeout is one day
#'
#' @examples
#' \dontrun{
#' waitForNodesToComplete(poolId = "testCluster", timeout = 3600)
#' }
#' @export
waitForNodesToComplete <- function(poolId, timeout = 86400) {
  cat("Booting compute nodes. . . ", fill = TRUE)
  config <- getConfiguration()
  pool <- config$batchClient$poolOperations$getPool(poolId)

  # Validate the getPool request first, before setting the progress bar
  if (!is.null(pool$code) && !is.null(pool$message)) {
    stop(sprintf("Code: %s - Message: %s", pool$code, pool$message))
  }

  if (pool$targetDedicatedNodes + pool$targetLowPriorityNodes < 0) {
    stop("Pool count needs to be greater or equal to 0.")
  }

  # Case: When user sets their pools to autoscale from 0 nodes
  if (pool$targetDedicatedNodes + pool$targetLowPriorityNodes == 0) {
    return(0)
  }

  totalNodes <-
    pool$targetDedicatedNodes + pool$targetLowPriorityNodes

  pb <-
    txtProgressBar(min = 0,
                   max = totalNodes,
                   style = 3)

  timeToTimeout <- Sys.time() + timeout

  while (Sys.time() < timeToTimeout) {
    pool <- config$batchClient$poolOperations$getPool(poolId)

    if (!is.null(pool$resizeErrors)) {
      cat("\n")

      resizeErrors <- ""
      for (i in 1:length(pool$resizeErrors)) {
        resizeErrors <-
          paste0(
            resizeErrors,
            sprintf(
              "Code: %s - Message: %s \n",
              pool$resizeErrors[[i]]$code,
              pool$resizeErrors[[i]]$message
            )
          )
      }

      stop(resizeErrors)
    }

    nodes <- pool <- config$batchClient$poolOperations$listPoolNodes(
      poolId)

    if (!is.null(nodes$value) && length(nodes$value) > 0) {
      nodesInfo <- .processNodeCount(nodes)

      currentProgressBarCount <- nodesInfo$currentNodeCount
      nodesWithFailures <- nodesInfo$nodesWithFailures

      if (currentProgressBarCount >= pb$getVal()) {
        setTxtProgressBar(pb, currentProgressBarCount)
      }

      .showNodesFailure(nodesWithFailures)
    }

    if (pb$getVal() >= totalNodes) {
      cat("\n")
      return(0)
    }

    Sys.sleep(30)
  }

  rAzureBatch::deletePool(poolId)
  stop("Timeout expired")
}

.processNodeCount <- function(nodes) {
  nodesWithFailures <- c()
  currentNodeCount <- 0
  nodesState <- list(
    idle = as.integer(0),
    creating = as.integer(0),
    starting = as.integer(0),
    waitingforstarttask = as.integer(0),
    starttaskfailed = as.integer(0),
    preempted = as.integer(0),
    running = as.integer(0),
    other = as.integer(0)
  )

  for (i in 1:length(nodes$value)) {
    state <- nodes$value[[i]]$state
    if (is.null(nodesState[[state]])) {
      nodesState[["other"]] <- nodesState[["other"]] + 1
    } else {
      nodesState[[state]] <- nodesState[[state]] + as.integer(1)
    }

    # The progress total count is the number of the nodes. Each node counts as 1.
    # If a node is not in idle, prempted, running, or start task failed, the value is
    # less than 1. The default value is 0 because the node has not been allocated to
    # the pool yet.
    nodeValue <- switch(
      nodes$value[[i]]$state,
      "idle" = {
        1
      },
      "creating" = {
        0.25
      },
      "starting" = {
        0.50
      },
      "waitingforstarttask" = {
        0.75
      },
      "starttaskfailed" = {
        nodesWithFailures <- c(nodesWithFailures, nodes$value[[i]]$id)
        1
      },
      "preempted" = {
        1
      },
      "running" = {
        1
      },
      0
    )

    currentNodeCount <-
      currentNodeCount + nodeValue
  }
  return(list(currentNodeCount = currentNodeCount, nodesWithFailures = nodesWithFailures, nodesState = nodesState))
}

.showNodesFailure <- function(nodesWithFailures) {
  if (length(nodesWithFailures) > 0) {
    nodesFailureWarningLabel <-
      sprintf(
        "The following %i nodes failed while running the start task:\n",
        length(nodesWithFailures)
      )
    for (i in 1:length(nodesWithFailures)) {
      nodesFailureWarningLabel <-
        paste0(nodesFailureWarningLabel,
               sprintf("%s\n", nodesWithFailures[i]))
    }

    warning(nodesFailureWarningLabel)
  }
}

#' Utility function for creating an output file
#'
#' @param filePattern a pattern indicating which file(s) to upload
#' @param url the destination blob or virtual directory within the Azure Storage container
#'
#' @export
createOutputFile <- function(filePattern, url) {
  output <- list(
    filePattern = filePattern,
    destination = list(container = list(containerUrl = url)),
    uploadOptions = list(uploadCondition = "taskCompletion")
  )

  # Parsing url to obtain container's virtual directory path
  # sample url: "https://accountname.blob.core.windows.net/outputs?se=2017-07-31&sr=c&st=2017-07-12"
  # after split by "/"
  # parsedValue[1] is "https"
  # parsedValue[2] is ""
  # parsedValue[3] is "accountname.blob.core.windows.net"
  # parsedValue[4] is "outputs?se=2017-07-31&sr=c&st=2017-07-12"
  parsedValue <- strsplit(url, "/")[[1]]

  baseUrl <- paste0(parsedValue[1], "//", parsedValue[3])
  urlPath <- sub(baseUrl, "", url)
  parsedUrlPath <- strsplit(urlPath, "?", fixed = TRUE)[[1]]

  storageContainerPath <- parsedUrlPath[1]
  queryParameters <- parsedUrlPath[2]
  virtualDirectory <-
    strsplit(substring(storageContainerPath, 2, nchar(storageContainerPath)), "/", fixed = TRUE)

  containerName <- virtualDirectory[[1]][1]
  containerUrl <-
    paste0(baseUrl, "/", containerName, "?", queryParameters)

  # Verify directory has multiple directories
  if (length(virtualDirectory[[1]]) > 1) {
    # Rebuilding output path for the file upload
    path <- ""
    for (i in 2:length(virtualDirectory[[1]])) {
      path <- paste0(path, virtualDirectory[[1]][i], "/")
    }

    path <- substring(path, 1, nchar(path) - 1)
    output$destination$container$path <- path
  }

  output$destination$container$containerUrl <- containerUrl
  output
}

getXmlValues <- function(xmlResponse, xmlPath) {
  xml2::xml_text(xml2::xml_find_all(xmlResponse, xmlPath))
}

saveMetadataBlob <- function(jobId, metadata) {
  xmlNode <- "<metadata>"
  if (length(metadata) > 0) {
    for (i in 1:length(metadata)) {
      xmlNode <-
        paste0(
          xmlNode,
          sprintf(
            "<%s>%s</%s>",
            metadata[[i]]$name,
            metadata[[i]]$value,
            metadata[[i]]$name
          )
        )
    }
  }
  xmlNode <- paste0(xmlNode, "</metadata>")
  saveXmlBlob(jobId, xmlNode, "metadata")
}

saveXmlBlob <- function(jobId, xmlBlock, name) {
  config <- getConfiguration()
  storageClient <- config$storageClient

  xmlFile <- paste0(jobId, "-", name, ".rds")
  saveRDS(xmlBlock, file = xmlFile)
  storageClient$blobOperations$uploadBlob(jobId, paste0(getwd(), "/", xmlFile))
  file.remove(xmlFile)
}

readMetadataBlob <- function(jobId) {
  config <- getConfiguration()
  storageClient <- config$storageClient

  tempFile <- tempfile(paste0(jobId, "-metadata"), fileext = ".rds")
  result <- storageClient$blobOperations$downloadBlob(
    jobId,
    paste0(jobId, "-metadata.rds"),
    downloadPath = tempFile,
    overwrite = TRUE
  )

  if (is.vector(result)) {
    result <- readRDS(tempFile)
    result <- xml2::read_xml(result)
    chunkSize <- getXmlValues(result, ".//chunkSize")
    packages <- getXmlValues(result, ".//packages")
    errorHandling <- getXmlValues(result, ".//errorHandling")
    wait <- getXmlValues(result, ".//wait")
    enableCloudCombine <-
      getXmlValues(result, ".//enableCloudCombine")

    metadata <-
      list(
        chunkSize = chunkSize,
        packages = packages,
        errorHandling = errorHandling,
        enableCloudCombine = enableCloudCombine,
        wait = wait
      )

    return(metadata)
  } else {
    return(NULL)
  }
}

hasDataSet <- function(list) {
  if (length(list) > 0) {
    for (arg in list[[1]]) {
      # Data frames are shown as list in the foreach iterator
      if (typeof(arg) == "list") {
        return(TRUE)
      }
    }
  }

  return(FALSE)
}

getHttpErrorMessage <- function(responseObj) {
  detailMessage <- paste0(responseObj$code, ": ", responseObj$message$value)

  if (length(responseObj$values) > 0) {
    for (i in 1:length(responseObj$values)) {
      detailMessage <- paste0(detailMessage, "\r\n", responseObj$values[[i]]$key, ": ", responseObj$values[[i]]$value)
    }
  }
  detailMessage <- paste0(detailMessage, "\r\nodata.metadata: ", responseObj$odata.metadata)
  return(detailMessage)
}

viewErrors <- function(id, failTasks) {
  config <- getConfiguration()
  storageClient <- config$storageClient
  
  sasToken <- storageClient$generateSasToken("r", "c", id)
  queryParameterUrl <- "?"
  
  for (query in names(sasToken)) {
    queryParameterUrl <-
      paste0(queryParameterUrl,
             query,
             "=",
             RCurl::curlEscape(sasToken[[query]]),
             "&")
  }
  
  queryParameterUrl <-
    substr(queryParameterUrl, 1, nchar(queryParameterUrl) - 1)
  
  tempDir <- tempfile()
  dir.create(tempDir)
  htmlFile <- file.path(tempDir, paste0(id, ".html"))
  azureStorageUrl <-
    paste0("http://",
           storageCredentials$name,
           sprintf(".blob.%s/", storageCredentials$endpointSuffix),
           id)
  
  staticHtml <- "<h1>Errors:</h1>"
  for (i in 1:length(failTasks)) {
    if (failTasks[i] == 1) {
      stdoutFile <- paste0(azureStorageUrl, "/", "stdout")
      stderrFile <- paste0(azureStorageUrl, "/", "stderr")
      rlogFile <- paste0(azureStorageUrl, "/", "logs")
      
      stdoutFile <-
        paste0(stdoutFile,
               "/",
               id,
               "-task",
               i,
               "-stdout.txt",
               queryParameterUrl)
      stderrFile <-
        paste0(stderrFile,
               "/",
               id,
               "-task",
               i,
               "-stderr.txt",
               queryParameterUrl)
      rlogFile <-
        paste0(rlogFile,
               "/",
               id,
               "-task",
               i,
               ".txt",
               queryParameterUrl)
      
      staticHtml <-
        paste0(
          staticHtml,
          "Task ",
          i,
          " | <a href='",
          stdoutFile,
          "'>",
          "stdout.txt",
          "</a> |",
          " <a href='",
          stderrFile,
          "'>",
          "stderr.txt",
          "</a> | <a href='",
          rlogFile,
          "'>",
          "R output",
          "</a> <br/>"
        )
    }
  }
  
  write(staticHtml, htmlFile)
  
  viewer <- getOption("viewer")
  if (!is.null(viewer)) {
    viewer(htmlFile)
  }
}

